mantidimaging/core/io/saver.py

```diff
@@ -91,11 +93,25 @@ def save(images: Images,
     output_dir = os.path.abspath(os.path.expanduser(output_dir))
     make_dirs_if_needed(output_dir, overwrite_all)
 
+    # Do rescale if needed.
+    if pixel_depth is None or pixel_depth == "float32":
+        rescale_params = None
+    elif pixel_depth == "int16":
+        int16_size = 65536
+        max_value = images.data.max()
+        slope = max_value / int16_size
+        rescale_params = {"offset": 0, "slope": slope}
+
+        # Overwrite images with the copy that has been rescaled.
+        images = do_rescale(images, images.data.min(), max_value, int16_size - 1)
+    else:
+        raise ValueError("The pixel depth given is not handled: " + pixel_depth)
```

Comment:
I can't think of any way to avoid copying the data with this approach of rescaling the whole stack (which requires 1 copy), plus then changing the data type of the whole stack (which requires another copy, but that one is with half the size because 32->16bit).

I think the way to go here is to change the for loop [below](https://github.com/mantidproject/mantidimaging/blob/465_SavingToDifferentGitLevels/mantidimaging/core/io/saver.py#L140-L144) to rescale a copy of the image (e.g. `write_func(RescaleFilter.do_rescale(np.copy(data[idx])...)`. This should also check if a rescale is necessary at all (i.e. dtypes differ)

The easiest way would be to extract the rescaling code from `RescaleFilter` into another static func that we call with the `data[idx]`, as well as within the filter itself

Reply:
Done :+1: